refactor: integrate AVS, operator and dogfood modules

trestinlsd opened the pull request:
## Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

----

Add epoch hook implementation for operator
Mockavs have been replaced with actual avs



<!--
< < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.

PR review checkboxes:

I have...

- [ ] added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] included the correct
      [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json)
      in the PR title
- [ ] targeted the correct branch
      (see [PR Targeting](https://github.com/ExocoreNetwork/exocore/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link in the PR description to the relevant issue or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all required CI checks have passed

Code maintenance:

I have...

- [ ] written unit and integration [tests](https://github.com/ExocoreNetwork/exocore/blob/main/CONTRIBUTING.md#testing)
- [ ] added relevant [`godoc`](https://go.dev/blog/godoc) and [code comments](https://blog.jbowen.dev/2019/09/the-magic-of-go-comments/).
- [ ] updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)

______

### Reviewers Checklist

**All** items are required.
Please add a note if the item is not applicable
and please add your handle next to the items reviewed
if you only reviewed selected items.

I have...

- [ ] confirmed the correct
      [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json)
      in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] confirmed that this PR does not change production code

-->

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit


- **New Features**
  - Introduced a structured type for managing AVS registration and deregistration parameters.
  - Added functions for generating AVS addresses and retrieving chain ID without revisions.
  - Implemented a new CLI command for querying AVS addresses by chain ID.

- **Improvements**
  - Enhanced initialization logic for key components, ensuring proper dependency management.
  - Improved accuracy in obtaining AVS addresses tied to operators.
  - Streamlined genesis state initialization and validation process, potentially affecting data integrity checks.
  - Enhanced error handling and validation in operator opt-in/out processes.

- **Bug Fixes**
  - Adjusted functionality in the `EndBlock` method to remove outdated operations related to voting power updates.

- **Documentation**
  - Updated comments to clarify dependencies and functionalities across various components.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

trestinlsd committed:
feat:Add epoch hook implementation for operator module

trestinlsd committed:
fix:Replace chainid with actual avs

mikebraver commented on x/operator/keeper/impl_epoch_hook.go:
```diff
@@ -0,0 +1,43 @@
+	for _, avs := range avsList {
+		err := wrapper.keeper.UpdateVotingPower(ctx, avs)
+		if err != nil {
+			panic(err)
```
Don't use panic here

trestinlsd replied:
fixed

trestinlsd committed:
fix:fix using panic for error handling

leonz789 commented on app/app.go:
```diff
@@ -584,6 +573,7 @@ func NewExocoreApp(
 			app.StakingKeeper.EpochsHooks(), // at this point, the order is irrelevant.
 			app.ExomintKeeper.EpochsHooks(), // however, this may change once we have distribution
 			app.AVSManagerKeeper.EpochsHooks(),
+			app.OperatorKeeper.EpochsHooks(),
```
EpochHooks wrapped the pointer of OperatorKeeper, and app.OperatorKeeper is initialized after this register with `NewOperatorKeerper`, so here in this hooks is will have a `nil` operatorKeeper

trestinlsd replied:
fixed

leonz789 commented on x/operator/keeper/consensus_keys.go:
```diff
@@ -352,12 +352,7 @@ func (k Keeper) GetActiveOperatorsForChainID(
-		return nil, nil
-	}
-
+	avsAddr := k.avsKeeper.GetAVSAddrByChainID(ctx, chainID)
```
check case "" ?

trestinlsd replied:
fixed

trestinlsd committed:
fix:fix epoch hooks nil

MaxMustermann2 committed:
major refactor: avs, dogfood, operator

- Operator: Use new wrapped consensus key object for convenience
- Avs: Hex addresses are first class citizens, except in the case of
  operators
- Dogfood: Associated changes to keep up with operator, and reorder
  genesis handling to be more inline with logic

MaxMustermann2 committed:
Merge remote-tracking branch 'devin/develop' into pr-149

MaxMustermann2 committed:
fix(local): update for new genesis format

MaxMustermann2 committed:
Merge remote-tracking branch 'upstream/develop' into pr-149

MaxMustermann2 committed:
refactor(avs): use IERC165 for blank contract

MaxMustermann2 commented on x/operator/keeper/usd_value.go:
```diff
@@ -228,10 +225,7 @@ func (k Keeper) GetVotePowerForChainID(
-	if err != nil {
-		return 0, err
-	}
+	avsAddr := k.avsKeeper.GetAVSAddrByChainID(ctx, chainID)
```
It can not return an error.

MaxMustermann2 commented on x/dogfood/keeper/genesis.go:
```diff
@@ -34,27 +34,36 @@ func (k Keeper) InitGenesis(
+	var avsAddr string
+	var err error
+	// the avs module will remove the revision by itself
+	if avsAddr, err = k.avsKeeper.RegisterAVSWithChainID(ctx, &avstypes.AVSRegisterOrDeregisterParams{
+		AvsName:           ctx.ChainID(),
+		AssetID:           genState.Params.AssetIDs,
+		UnbondingPeriod:   uint64(genState.Params.EpochsUntilUnbonded),
+		MinSelfDelegation: genState.Params.MinSelfDelegation.Uint64(),
+		EpochIdentifier:   epochID,
+		ChainID:           ctx.ChainID(),
+	}); err != nil {
+		panic(fmt.Errorf("could not create the dogfood AVS: %s", err))
+	}
```
It is already handled.

MaxMustermann2 commented on x/operator/keeper/opt.go:
```diff
@@ -56,24 +61,29 @@
 		return delegationtypes.ErrOperatorNotExist
 	}
+	// check that the AVS is registered
+	if isAvs, _ := k.avsKeeper.IsAVS(ctx, avsAddr); !isAvs {
+		return types.ErrNoSuchAvs.Wrapf("AVS not found %s", avsAddr)
```
The error is not relevant. The `IsAVS` function returns `false, error` or `true, nil`.

MaxMustermann2 commented on x/operator/keeper/opt.go:
```diff
@@ -16,18 +14,25 @@
+		return delegationtypes.ErrOperatorNotExist
+	}
+	// check that the AVS is registered
+	if isAvs, _ := k.avsKeeper.IsAVS(ctx, avsAddr); !isAvs {
+		return types.ErrNoSuchAvs.Wrapf("AVS not found %s", avsAddr)
```
The error is not relevant.

MaxMustermann2 commented on x/operator/keeper/opt.go:
```diff
@@ -56,24 +61,29 @@ func (k *Keeper) OptIn(ctx sdk.Context, operatorAddress sdk.AccAddress, avsAddr
+	if err != nil {
+		return err
+	}
+	chainID, _ := k.avsKeeper.GetChainIDByAVSAddr(ctx, avsAddr)
+	return k.SetOperatorConsKeyForChainID(ctx, operatorAddress, chainID, key)
```
The error is not relevant.

MaxMustermann2 commented on x/operator/keeper/genesis.go:
```diff
@@ -13,31 +13,6 @@ func (k Keeper) InitGenesis(ctx sdk.Context, state types.GenesisState) []abci.Va
-				panic(err)
-			}
-		}
-	}
 	return []abci.ValidatorUpdate{}
 }
```
The issue already exists.

MaxMustermann2 commented on x/operator/keeper/slash.go:
```diff
@@ -233,16 +224,12 @@ func (k *Keeper) SetJailedState(ctx sdk.Context, consAddr sdk.ConsAddress, chain
-		k.Logger(ctx).Error(err.Error(), chainID)
-		return
-	}
+	avsAddr := k.avsKeeper.GetAVSAddrByChainID(ctx, chainID)
```
Not needed since we validated it prior to this.

MaxMustermann2 committed:
fix(app): add back delegation hooks

MaxMustermann2 committed:
feat(avs): add query for chainID to avsAddr

MaxMustermann2 committed:
refactor(avs): clarify function names

There was some confusion about checking whether a chainID is an AVS or
not. It has now been simplified.

MaxMustermann2 committed:
fix: update for avs changes

MaxMustermann2 committed:
fix(test): make the tests pass

cloud8little commented:
Test Update, all the following test scenarios are working well. 

- Consensus continues after 1 epoch
- Vote power changes in response to delegation and undelegation
- Operator CLI opt-in is via AVS hex address and not chainID
- Operator/dogfood queries work as before but via AVS hex address (or chainID, without the upgrade revision number)

trestinlsd committed:
fix:Fix the issue of too many contract parameters that cannot be compiled

cloud8little commented:
queryAVSInfo use Bench32 address while registerAVS use the hex address to store. 

```
	// The AVS registration is done by the calling contract.
	avsParams.AvsAddress = contract.CallerAddress.String()
	avsParams.Action = avskeeper.RegisterAction
	// Finally, update the AVS information in the keeper.
	err = p.avsKeeper.AVSInfoUpdate(ctx, avsParams)
```

```
func QueryAVSInfo() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "AVSInfo query",
		Short: "AVSInfo query",
		Long:  "AVSInfo query for current registered AVS",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			_, err := sdk.AccAddressFromBech32(args[0])
			if err != nil {
				return xerrors.Errorf("invalid  address,err:%s", err.Error())
			}
			clientCtx, err := client.GetClientQueryContext(cmd)
			if err != nil {
				return err
			}

			queryClient := types.NewQueryClient(clientCtx)
			req := &types.QueryAVSInfoReq{
				AVSAddress: args[0],
			}
			res, err := queryClient.QueryAVSInfo(context.Background(), req)
			if err != nil {
				return err
			}
			return clientCtx.PrintProto(res)
		},
	}
	flags.AddQueryFlagsToCmd(cmd)
	return cmd
}
```

trestinlsd committed:
fix:check as must use the hex address

MaxMustermann2 committed:
test: add dogfood integration tests

MaxMustermann2 committed:
Merge upstream from devin

MaxMustermann2 committed:
chore: lint

MaxMustermann2 commented on x/dogfood/keeper/setup_test.go:
```diff
@@ -0,0 +1,28 @@
+	suite.Run(t, s)
+}
+
+func (suite *KeeperTestSuite) SetupTest() {
+	suite.DoSetupTest()
+	epochID := suite.App.StakingKeeper.GetEpochIdentifier(suite.Ctx)
+	epochInfo, _ := suite.App.EpochsKeeper.GetEpochInfo(suite.Ctx, epochID)
+	suite.EpochDuration = epochInfo.Duration + time.Nanosecond // extra buffer
```
Not needed since this check is performed at genesis

MaxMustermann2 commented on testutil/tx/signer.go:
```diff
@@ -39,6 +42,16 @@ func GenerateAddress() common.Address {
 	return addr
 }
 
+// GenerateConsensusKey generates a consensus key.
+func GenerateConsensusKey() operatortypes.WrappedConsKey {
+	privVal := mock.NewPV()
+	pubKey, err := privVal.GetPubKey()
+	if err != nil {
+		return nil
+	}
+	return operatortypes.NewWrappedConsKeyFromHex(hexutil.Encode(pubKey.Bytes()))
+}
```
Not needed, test only

cloud8little commented:
registerAVS via precompile contract failed.

```
I[2024-07-31|02:05:35.564] avsInfo:                                     is=<nil>
I[2024-07-31|02:05:35.565] avsInfo:                                     is="info:<name:\"avstesttest01\" avs_address:\"0xd1f4f71A831c8517E75e5A3F2FaE9bF32a1eAc1e\" min_stake_amount:100 task_addr:\"0xd59f558D8f2EfC2256764D6dEc389C0b60B1f488\" slash_addr:\"0xd59f558D8f2EfC2256764D6dEc389C0b60B1f488\" reward_addr:\"0xd59f558D8f2EfC2256764D6dEc389C0b60B1f488\" avs_owner_address:\"exo1j6qqrnw0w4vxzxcuqlzcf9ywglcqn3kkawd0gg\" asset_ids:\"0x10\" avs_unbonding_period:200 min_self_delegation:10 epoch_identifier:\"hour\" min_opt_in_operators:10 min_total_stake_amount:20 starting_epoch:22 avs_reward:\"300000000000000000\" avs_slash:\"400000000000000000\" > "
Failed to update AVS info the avsaddress is :0xd1f4f71A831c8517E75e5A3F2FaE9bF32a1eAc1e: Error: Already registered [ExocoreNetwork/exocore/x/avs/keeper/keeper.go:74]
2:05AM ERR call avs precompile error err="the avsaddress is :0xd1f4f71A831c8517E75e5A3F2FaE9bF32a1eAc1e: Error: Already registered" module="avs precompile"
I[2024-07-31|02:05:35.566] avsInfo:                                     is="info:<name:\"avstesttest01\" avs_address:\"0xd1f4f71A831c8517E75e5A3F2FaE9bF32a1eAc1e\" min_stake_amount:100 task_addr:\"0xd59f558D8f2EfC2256764D6dEc389C0b60B1f488\" slash_addr:\"0xd59f558D8f2EfC2256764D6dEc389C0b60B1f488\" reward_addr:\"0xd59f558D8f2EfC2256764D6dEc389C0b60B1f488\" avs_owner_address:\"exo1j6qqrnw0w4vxzxcuqlzcf9ywglcqn3kkawd0gg\" asset_ids:\"0x10\" avs_unbonding_period:200 min_self_delegation:10 epoch_identifier:\"hour\" min_opt_in_operators:10 min_total_stake_amount:20 starting_epoch:22 avs_reward:\"300000000000000000\" avs_slash:\"400000000000000000\" > "
Failed to update AVS info the avsaddress is :0xd1f4f71A831c8517E75e5A3F2FaE9bF32a1eAc1e: Error: Already registered [ExocoreNetwork/exocore/x/avs/keeper/keeper.go:74]
```

cloud8little commented:
> registerAVS via precompile contract failed.
> 
> ```
> I[2024-07-31|02:05:35.564] avsInfo:                                     is=<nil>
> I[2024-07-31|02:05:35.565] avsInfo:                                     is="info:<name:\"avstesttest01\" avs_address:\"0xd1f4f71A831c8517E75e5A3F2FaE9bF32a1eAc1e\" min_stake_amount:100 task_addr:\"0xd59f558D8f2EfC2256764D6dEc389C0b60B1f488\" slash_addr:\"0xd59f558D8f2EfC2256764D6dEc389C0b60B1f488\" reward_addr:\"0xd59f558D8f2EfC2256764D6dEc389C0b60B1f488\" avs_owner_address:\"exo1j6qqrnw0w4vxzxcuqlzcf9ywglcqn3kkawd0gg\" asset_ids:\"0x10\" avs_unbonding_period:200 min_self_delegation:10 epoch_identifier:\"hour\" min_opt_in_operators:10 min_total_stake_amount:20 starting_epoch:22 avs_reward:\"300000000000000000\" avs_slash:\"400000000000000000\" > "
> Failed to update AVS info the avsaddress is :0xd1f4f71A831c8517E75e5A3F2FaE9bF32a1eAc1e: Error: Already registered [ExocoreNetwork/exocore/x/avs/keeper/keeper.go:74]
> 2:05AM ERR call avs precompile error err="the avsaddress is :0xd1f4f71A831c8517E75e5A3F2FaE9bF32a1eAc1e: Error: Already registered" module="avs precompile"
> I[2024-07-31|02:05:35.566] avsInfo:                                     is="info:<name:\"avstesttest01\" avs_address:\"0xd1f4f71A831c8517E75e5A3F2FaE9bF32a1eAc1e\" min_stake_amount:100 task_addr:\"0xd59f558D8f2EfC2256764D6dEc389C0b60B1f488\" slash_addr:\"0xd59f558D8f2EfC2256764D6dEc389C0b60B1f488\" reward_addr:\"0xd59f558D8f2EfC2256764D6dEc389C0b60B1f488\" avs_owner_address:\"exo1j6qqrnw0w4vxzxcuqlzcf9ywglcqn3kkawd0gg\" asset_ids:\"0x10\" avs_unbonding_period:200 min_self_delegation:10 epoch_identifier:\"hour\" min_opt_in_operators:10 min_total_stake_amount:20 starting_epoch:22 avs_reward:\"300000000000000000\" avs_slash:\"400000000000000000\" > "
> Failed to update AVS info the avsaddress is :0xd1f4f71A831c8517E75e5A3F2FaE9bF32a1eAc1e: Error: Already registered [ExocoreNetwork/exocore/x/avs/keeper/keeper.go:74]
> ```

This is because registerAVS tx costs more than `10000000 `gas while the max gas limit is `10000000  `in local_node shell script, increase it to `100000000`, registerAVS, updateAVS, query avs Info all test passed.
